lxd/proxy_device_utils.go

```diff
@@ -0,0 +1,93 @@
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"syscall"
+)
+
+type proxyProcInfo struct {
+	listenPid   string
+	connectPid  string
+	connectAddr string
+	listenAddr  string
+}
+
+func createProxyDevInfoFile(devicesPath string, proxyDev string, proxyPid int) error {
+	devFileName := fmt.Sprintf("proxy.%s", proxyDev)
+	filePath := filepath.Join(devicesPath, devFileName)
+	f, err := os.Create(filePath)
+
+	if err != nil {
+		return err
+	}
+
+	defer f.Close()
+
+	info := fmt.Sprintf("%d", proxyPid)
+	_, err = f.WriteString(info)
+
+	return err
+}
+
+func setupProxyProcInfo(c container, device map[string]string) (*proxyProcInfo, error) {
+	pid := c.InitPID()
+	containerPid := strconv.Itoa(int(pid))
+	lxdPid := strconv.Itoa(os.Getpid())
+
+	connectAddr := device["connect"]
+	listenAddr := device["listen"]
+
+	connectionType := strings.SplitN(connectAddr, ":", 2)[0]
+	listenerType := strings.SplitN(listenAddr, ":", 2)[0]
+
+	if connectionType != "tcp" {
+		return nil, fmt.Errorf("Proxy device currently doesnt support the connection type: %s", connectionType)
+	}
+	if listenerType != "tcp" {
+		return nil, fmt.Errorf("Proxy device currently doesnt support the listener type: %s", listenerType)
+	}
+
+	listenPid := "-1"
+	connectPid := "-1"
+
+	if device["bind"] == "container" {
+		listenPid = containerPid
+		connectPid = lxdPid
+	} else if device["bind"] == "host" {
+		listenPid = lxdPid
+		connectPid = containerPid
+	} else {
+		return nil, fmt.Errorf("No indicated binding side")
+	}
+
+	p := &proxyProcInfo{
+		listenPid:   listenPid,
+		connectPid:  connectPid,
+		connectAddr: connectAddr,
+		listenAddr:  listenAddr,
+	}
+
+	return p, nil
+}
+
+func killProxyProc(devPath string) error {
+	contents, err := ioutil.ReadFile(devPath)
+	if err != nil {
+		return err
+	}
+
+	pid, _ := strconv.Atoi(string(contents))
+	if err != nil {
+		return err
+	}
+
+	syscall.Kill(pid, syscall.SIGINT)
```

Comment:
That should be SIGTERM

Reply:
You should also perform a few more checks in this function:
 - Make sure that the process exist
 - Make sure that the process is what it should be
 - Add a waitloop to check that the process is gone after SIGTERM and if not gone after a few seconds, then use SIGKILL

networkKillDnsmasq should be a pretty close example of this